ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -59,18 +68,18 @@ noProgressReporting = return $ ProgressReporting
 data State
     = NotStarted
     | Stopped
-    | Running (Async ())
+    | Running (Barrier ())
 
 -- | State transitions used in 'delayedProgressReporting'
 data Transition = Event ProgressEvent | StopProgress
 
-updateState :: IO (Async ()) -> Transition -> State -> IO State
+updateState :: Barrier () -> Transition -> State -> IO State
 updateState _      _                    Stopped     = pure Stopped
-updateState start (Event KickStarted)   NotStarted  = Running <$> start
-updateState start (Event KickStarted)   (Running a) = cancel a >> Running <$> start
-updateState _     (Event KickCompleted) (Running a) = cancel a $> NotStarted
+updateState start (Event KickStarted)   NotStarted  = pure $ Running start
+updateState start (Event KickStarted)   (Running a) = signalBarrier a () $> Running start
```

Comment:
Hmmm. Does this work? We signal the barrier and then go back into the running state, but won't the barrier have cancelled the whole progress reporting??

Reply:
we cancel the previous job and leave the new job cancellation here.